packages/@aws-cdk/aws-s3-deployment/lambda/test/aws

```diff
@@ -24,4 +24,4 @@ if sys.argv[2] == "sync":
     sys.argv[4] = "contents.zip"
 
 with open("./aws.out", "a") as myfile:
-    myfile.write(" ".join(sys.argv[1:]) + "\n")
+    myfile.write(json.dumps(sys.argv[1:]) + "\n")
```

Comment:
Im not sure about this. It makes the test feel less natural. 

When we assert the `aws` commands, its easier looking at them as a single string, as that is what they eventually are. The fact that string is being generated from an array seems like something the test shouldn't be aware of. Think about how you would test it if we were to actually invoke `aws` cli and inspect its output?

It also has the benefit of making the commands copy-pastable. 

Also, `json.dumps` will recursively dump every element in the array, possible masquerading some bugs or making it harder to debug. The fact we are simply `joining`, makes it clear that we are only expecting strings to be in that array. Basically, `sys.argv` is an array of strings, therefore shouldn't be treated as a generic `dumpable` object.

Reply:
We are using `subprocess.check_call(["python3", aws] + list(args))` eventually to execute the aws cli command. This is simply a way of serializing the list of strings that were passed to the cli.

Reply:
I understand. But i'm not sure how your response addresses my concerns regarding the affects of this on the tests...It seems like you could have left this as is without the need to change all the tests. So i'm not sure why this is an improvement.

Reply:
Yes the tests will pass, which is exactly the issue. If you keep it how it was previously, this: `["arg1", "arg2 has space"]` turns into `arg1 arg2 has space`. You lose information.

Reply:
You can also provide another function to serialize and deserialize a list, I have no preference for json - it is just convenient.

Reply:
Gotcha. So Basically, we want to validate that we are invoking `aws` like so:

`s3 sync --delete /tmp/temp s3://temp/ "--cache-control" "max-age=600"`, 

And not like so:

`s3 sync --delete /tmp/temp s3://temp/ "--cache-control max-age=600"`

Which indeed causes the `Unknown options: --cache-control 'max-age=600'` error. 

And the only way to do that is to properly serialize them, as otherwise they are both asserted as:

`s3 sync --delete /tmp/temp s3://temp/ --cache-control max-age=600`

This means that its not just a bug with `cache-control`, but essentially every `SystemMetadata` that is passed will cause this error, which was also the original title. Right? 

If so, let's change the title back and also comment in the issue that the problem is more severe than initially described.

Reply:
Hmm, not sure why the title was changed. Also my description already explains the problem?

Reply:
It was changed by @eladb to be more specific and describe the actual problem. This was before understanding the full implications of this. 

In any case, the description doesn't explain the problem, it explains the root-cause. Also, it specifically mentions `cache-control`, even though the problem is not directly related to it. 

As you might know, the PR title is displayed in our [release notes](https://github.com/aws/aws-cdk/releases). For customers to understand exactly what they can expect in the new version, the title has to describe the problems (symptoms), and not the solution or the root-cause. 

For this case, I would go with:

`fix(s3-deployment): passing any system metadata causes lambda to fail on "Unknown options:" when invoking aws cli`. 

We prefer descriptiveness over brevity.

In addition, users who land on the original issue, should be aware that it has greater implications that what was originally posted. As it may apply to them. So we should add a comment that clarifies this.

Makes sense?

Also, please don't mark a conversation as `resolved` if there are still open questions in it, it causes the discussion to be hidden in the review.

Reply:
title changed

Reply:
👍